articles/implementing-caching-in-nodejs-using-redis/index.md

```diff
@@ -0,0 +1,257 @@
+- Have [Postman](https://www.postman.com/) installed on your computer.
+- Have a Linux or macOS based operating system.
+
+### Application overview.
+In this article, we will walk through the steps of implementing caching in a Node.js app using [redis](https://redis.io/). We will implement a REST API using the Express.js framework. We will be making requests to [recipe labs api](http://www.recipepuppy.com/?ref=hackernoon.com). It is a public API for accessing recipes of different food items.
+
+#### Setting up the application.
+Having installed [Node.js](https://nodejs.org/en/), we are equipped with [npm](https://www.npmjs.com/) which we will use to install the necessary dependencies. Proceed to the directory where you want the project to be and run the following command.
+
+```bash
+npm init --yes
+```
+
+With the `package.json` file generated, we can continue to the next step.
+
+#### Installing dependencies.
+We will need the following dependencies:
+
+- **axios**: For handling requests to the public API.
+- **express**: For setting up a working Express.js environment.
+- **nodemon**: For automatically restarting the server whenever we make changes.
+- **redis**: For implementing caching.
+
+We install them using the following command:
+
+```bash
+npm install axios redis express nodemon
+```
+
+#### Configuring package.json
+To incorporate `nodemon` in our development, we make the following adjustments in the `package.json` file:
+
+```javascript
+    {
+        "scripts" : {
+            //add the following
+            "dev":"nodemon app.js",
+            "start":"node app.js"
+        }
+    }
+```
+
+#### Creating Express.js app.
+In the same directory, we create an `app.js` file. The file should be as below:
+
+```javascript
+//import express
+const express = require('express');
+//import axios
+const axios = require('axios');
+//import redis
+const redis = require('redis');
+
+//create an instance of an app.
+const app = express();
+
+//configure the port
+const PORT = 3001;
+
+//start the server
+app.listen(PORT, () => {
+    console.log(`app started on ${PORT}`)
+});
+```
+
+#### Fetching data from the public API.
+Since we have set up the application, we can now fetch the data from the public API. To do this we need to implement a route. We shall fetch and return the data on the route.
+
+The route shall be as follows:
+```javascript
+app.get(`/recipe/:foodItem`, async (req,res,next) => {
+
+    try {
+
+        //destructure the foodItem from params.
+        let {foodItem} = req.params;
+
+        //fetch the data.
+        const recipe = await axios.get(`http://www.recipepuppy.com/api/?q=${fooditem}`).catch(console.log);
+
+        //return a response.
+        return res.send({
+            success:true,
+            message:recipe.data.results
+        });
+
+    }catch(err){
+
+        return res.send({
+            success:false,
+            message:err
+        });
+
+    };    
+
+});
+```
```

Comment:
This route does not work.
I added a simple `hello world` route to test if the app itself was functional and it worked but the `/recipe/:foodItem` route doesn't work.
```Javascript
//import express
const express = require("express");
//import axios
const axios = require("axios");
//import redis
const redis = require("redis");

//create an instance of an app.
const app = express();

//configure the port
const PORT = 3001;

//start the server
app.listen(PORT, () => {
  console.log(`app started on ${PORT}`);
});

app.get("/", (req, res) => {
  res.send("Hello World!");
});

app.get(`/recipe/:foodItem`, async (req, res, next) => {
  try {
    //destructure the foodItem from params.
    let { foodItem } = req.params;

    //fetch the data.
    const recipe = await axios
      .get(`http://www.recipepuppy.com/api/?q=${fooditem}`)
      .catch(console.log);

    //return a response.
    return res.send({
      success: true,
      message: recipe.data.results,
    });
  } catch (err) {
    return res.send({
      success: false,
      message: err,
    });
  }
});
```

Attached below is my code currently and a screenshot of my postman request for the "hello world route

![2021-01-27](https://user-images.githubusercontent.com/47227271/105978038-244e9e80-60a3-11eb-84f8-bcc43a9d9e7c.png)

Attached below is the postman request for the "/recipe/:foodItem" request

![image](https://user-images.githubusercontent.com/47227271/105978511-af2f9900-60a3-11eb-9a21-a8fe421b1427.png)

Reply:
fixed